packages/gatsby/src/utils/cache.ts

```diff
@@ -1,58 +1,75 @@
-const fs = require(`fs-extra`)
-const manager = require(`cache-manager`)
-const fsStore = require(`cache-manager-fs-hash`)
-const path = require(`path`)
+import manager, { Store, StoreConfig, CachingConfig } from "cache-manager"
+import fs from "fs-extra"
+import fsStore from "cache-manager-fs-hash"
+import path from "path"
 
 const MAX_CACHE_SIZE = 250
 const TTL = Number.MAX_SAFE_INTEGER
 
-class Cache {
-  constructor({ name = `db`, store = fsStore } = {}) {
+interface ICacheProperties {
+  name?: string
+  store?: Store
+}
+
+export default class Cache {
+  public name: string
+  public store: Store
+  public cache?: manager.Cache
+
+  constructor({ name = `db`, store = fsStore }: ICacheProperties = {}) {
     this.name = name
     this.store = store
   }
 
-  get directory() {
+  get directory(): string {
     return path.join(process.cwd(), `.cache/caches/${this.name}`)
   }
 
-  init() {
+  init(): Cache {
     fs.ensureDirSync(this.directory)
 
-    const caches = [
+    const configs: StoreConfig[] = [
       {
         store: `memory`,
         max: MAX_CACHE_SIZE,
+        ttl: TTL,
       },
       {
         store: this.store,
+        ttl: TTL,
         options: {
           path: this.directory,
           ttl: TTL,
         },
       },
-    ].map(cache => manager.caching(cache))
+    ]
+
+    const caches = configs.map(cache => manager.caching(cache))
 
     this.cache = manager.multiCaching(caches)
 
     return this
   }
 
-  get(key) {
+  get<T = unknown>(key): Promise<T | undefined> {
     return new Promise(resolve => {
-      this.cache.get(key, (err, res) => {
+      // eslint-disable-next-line no-unused-expressions
+      this.cache?.get<T>(key, (err, res) => {
         resolve(err ? undefined : res)
       })
     })
   }
 
-  set(key, value, args = {}) {
+  set<T>(
```

Comment:
Shouldn't `T` be bound to the type of the cache? So wouldn't that mean you need a type parameter on the class that gets referenced here, rather than just this method? `new <Foo>Cache().get()`

Right now I think you need to do `new Cache().get<Foo>()` or something? But then you could basically pass on anything you'd want. Or am I missing something?

Reply:
Will the cache always have the same types?

Reply:
(i.e. will the items in the cache all be the same type)

Reply:
Unless I'm missing something, one cache could contain lots of different types of cache entry. This would mean we can't type the cache, and the best we can do is make the getter generic, and default to unknown, and let the user declare the type in the get call. e.g. `const myObject = cache.get<MyObjectType>("myKey")`. We could default to any rather than unknown if we want to be more lax about it.

Reply:
I looked through the codebase to better understand this cache. So from what I understand, multiple cache's will exist and be held in a map which can be looked up via a key. These cache's are created and used in plugins. From usages within our own codebase I've seen strings stored, booleans, objects. So I think using a generic here is correct. 

The only way to rethink this is to find a system that doesn't use a generic cache, but instead uses specific cache's for specific jobs that can have defined specs. I don't know if that is possible for our ecosystem or not. But it's for sure out of scope for this here.

Reply:
Sorry, that wasn't quite what this discussion was bout. lol my bad

Reply:
It's kind've what it was about! It does seem that they're used for any old thing

Reply:
If one cache instance holds multiple types of objects .... that's a code smell and I wouldn't expect there to be one for caching. Regardless, that's not to hold against the typing effort.

Reply:
I'm not sure how often that happens in practice, but there's nothing preventing it. Each plugin gets one cache, so it could certainly have a need to cache different types of object.